go/libkb/leveldb_test.go

```diff
@@ -0,0 +1,342 @@
+// Copyright 2016 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package libkb
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+	"time"
+)
+
+type teardowner struct {
+	sync.Mutex
+
+	actions  []func()
+	torndown bool
+}
+
+func (td *teardowner) register(teardownAction func()) {
+	td.Lock()
+	defer td.Unlock()
+	if td.torndown {
+		panic("already torndown")
+	}
+	td.actions = append(td.actions, teardownAction)
+}
+
+func (td *teardowner) teardown() {
+	td.Lock()
+	defer td.Unlock()
+	if td.torndown {
+		panic("already torndown")
+	}
+	for _, a := range td.actions {
+		a()
+	}
+}
+
+func createTempLevelDbForTest(tc *TestContext, td *teardowner) (*LevelDb, error) {
+	dir, err := ioutil.TempDir("", "level-db-test-")
+	if err != nil {
+		return nil, err
+	}
+
+	db := NewLevelDb(tc.G, func() string {
+		return filepath.Join(dir, "test.leveldb")
+	})
+
+	td.register(func() {
+		db.Close()
+		os.RemoveAll(dir)
+	})
+
+	return db, nil
+}
+
+func testLevelDbPut(db *LevelDb) (key DbKey, err error) {
+	key = DbKey{Key: "test-key", Typ: 0}
+	v := []byte{1, 2, 3, 4}
+	if err := db.Put(key, nil, v); err != nil {
+		return DbKey{}, err
+	}
+	if val, found, err := db.Get(key); err != nil {
+		return DbKey{}, err
+	} else if !found {
+		return DbKey{}, fmt.Errorf("stored object was not found by Get")
+	} else if !bytes.Equal(val, v) {
+		return DbKey{}, fmt.Errorf("stored object has incorrect data. expect %v, got %v", v, val)
+	}
+
+	return key, nil
+}
+
+func TestLevelDb(t *testing.T) {
+	var td teardowner
+
+	tests := []struct {
+		name     string
+		testBody func(t *testing.T)
+	}{
+		{
+			name: "simple", testBody: func(t *testing.T) {
+				tc := SetupTest(t, "LevelDb-simple", 0)
+				db, err := createTempLevelDbForTest(&tc, &td)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				key, err := testLevelDbPut(db)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				if err = db.Delete(key); err != nil {
+					t.Fatal(err)
+				}
+				_, found, err := db.Get(key)
+				if err != nil {
+					t.Fatal(err)
+				}
+				if found {
+					t.Fatalf("delete did not delete object")
+				}
+			},
+		},
+		{
+			name: "concurrent", testBody: func(t *testing.T) {
+				tc := SetupTest(t, "LevelDb-concurrent", 0)
+				db, err := createTempLevelDbForTest(&tc, &td)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				var wg sync.WaitGroup
+				wg.Add(2)
+				// synchronize between two doWhileOpenAndNukeIfCorrupted calls to know
+				// for sure they can happen concurrently.
+				ch := make(chan struct{})
+				go db.doWhileOpenAndNukeIfCorrupted(func() error {
```

Comment:
Ah I see. Yes. Maybe we could increase the timeouts to e.g. 8 seconds? That duration should be long enough for slow operation to complete, but any fast operation is not affected by how long that duration is.